src/main.rs

```diff
@@ -209,20 +202,102 @@ fn to_ca_certs(ca_certs_pem: &Vec<Vec<u8>>) -> Result<rustls::RootCertStore> {
     Ok(root_cert)
 }
 
-/// Initializes the tracing subscriber.
+/// Manages the log file and guards.
 ///
-/// If `log_dir` is `None` or the runtime is in debug mode, logs will be printed to stdout.
+/// `_guards` will flush the logs when they are dropped.
 ///
-/// Returns a vector of `WorkerGuard` that flushes the log when dropped.
-fn init_tracing(log_dir: Option<&Path>) -> Vec<WorkerGuard> {
-    let mut guards = vec![];
-    let subscriber = tracing_subscriber::Registry::default();
-    let file_name = format!("{}.log", env!("CARGO_PKG_NAME"));
+/// `dynamic_log_file_writer` wraps the log file to allow changing its path dynamically.
+/// If the log file is not provided, logs will be ignored by using `std::io::sink()`.
+struct LogManager {
+    _guards: Vec<WorkerGuard>,
+    dynamic_log_file_writer: DynamicLogFileWriter,
+}
+
+#[derive(Clone)]
+struct DynamicLogFileWriter {
+    writer: Arc<Mutex<Box<dyn Write + Send>>>,
+}
+
+impl DynamicLogFileWriter {
+    fn try_new(dir_path: Option<&Path>) -> Result<Self> {
+        Ok(Self {
+            writer: Arc::new(Mutex::new(DynamicLogFileWriter::create_writer(dir_path)?)),
+        })
+    }
+
+    fn create_log_file(dir_path: &Path) -> Result<File> {
+        if let Err(e) = create_dir_all(dir_path) {
+            bail!("Cannot create the directory recursively for {dir_path:?}: {e}");
+        }
+
+        let file_name = format!("{}.log", env!("CARGO_PKG_NAME"));
+
+        let file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(dir_path.join(file_name))
+            .map_err(|e| anyhow!("Cannot create the log file: {e}"));
 
-    let is_valid_file =
-        matches!(log_dir, Some(path) if std::fs::File::create(path.join(&file_name)).is_ok());
+        file
+    }
+
+    fn create_writer(log_dir: Option<&Path>) -> Result<Box<dyn Write + Send>> {
+        match log_dir {
+            Some(dir) => Ok(Box::new(DynamicLogFileWriter::create_log_file(dir)?)),
+            None => Ok(Box::new(std::io::sink())),
+        }
+    }
 
-    let stdout_layer = if !is_valid_file || cfg!(debug_assertions) {
+    fn change_log_dir(&mut self, old_dir: Option<&Path>, new_dir: Option<&Path>) -> Result<()> {
+        if old_dir.eq(&new_dir) {
+            info!("New log directory is the same as the old directory");
+            return Ok(());
+        }
+        if let Some(dir) = new_dir {
+            info!("Log directory will change to {}", dir.display());
+        }
+        let new_writer = DynamicLogFileWriter::create_writer(new_dir)?;
+        {
+            let mut old_writer = self.writer.lock().unwrap_or_else(PoisonError::into_inner);
```

Comment:
In short, there was no difference when I placed `old_writer.flush()` in that location.

1)
When `tracing-appender`'s `non_blocking` calls the `flush()`, 
the actual callee is `dynamic_log_file_writer` (or another wrapper struct), which wraps `old_writer`.
Thus, calling `old_writer.flush()` directly didn't immediately flush the logs.

Regarding the log loss issue you mentioned, the logs in memory will eventually be saved by `new_writer` even if the `old_writer` has already been dropped.

2)
The reason I added the code initially was to prevent the message "Log directory will change to {}" from being saved in the new directory. For example, if the new directory is `.../logs2`, it would be unhelpful to have "Log directory will change to .../logs2" within the logs2 directory itself, as it becomes self-referential.

3)
However, as I mentioned in 1), `old_writer.flush()` doesn't force an immediate flush, so I removed it.
Instead, I added a delay by inserting `let new_writer = DynamicLogFileWriter::create_writer(new_dir)?;` to ensure enough time for the log to be saved in the old directory.

Although it's not entirely deterministic, I believe this approach is reliable enough for changing the log directory. (manually tested)

Reply:
@danbi2990 
Thank you for the detailed explanation! 

I read your first response and now fully understand what you mentioned.
Looking into the implementation of `tracing-appender`'s `non_blocking`, I noticed that when `write()` is called, it sends logs to the worker thread via a channel. If the `is_lossy` option is enabled, it uses `try_send()`. Otherwise, it uses `send()`.
Here's the relevant code: [tracing-appender non_blocking.rs#L245](https://github.com/tokio-rs/tracing/blob/c6bedbe2725830c1e78cbcdb9168de69c98e42fc/tracing-appender/src/non_blocking.rs#L245)

I understand this aligns with what you explained.
Essentially, the `write()` function simply sends data to the channel, while flushing is fully determined by the recv loop in the worker thread.

What I wanted to highlight was the behavior of the operating system's file system buffer.
When looking into Rust's `std::unix::fs` implementation, it seems like `flush()` doesn’t actually do anything! 😂
Here's the relevant code: [Rust Unix fs.rs#L1286](https://github.com/rust-lang/rust/blob/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/sys/pal/unix/fs.rs#L1286)

As a result, due to the `tracing-appender`'s `non_blocking` implementation characteristics, the log remaining in the channel to the worker thread after the point when `log_dir` is changed will be written to the new `log_dir`.

Regarding points 2 and 3, I don't personally consider this a loss. However, if we were to treat this as a potential issue, we could fork the crate and add a `send(new_writer)` API. This way, the channel could be structured as [log, log, new_writer], maintaining the order of log generation and writer changes.
That said, I don’t see this as a loss, so I wouldn’t treat it as a problem!

After some additional research, I found that if I need to synchronize memory with the file system, I should use `sync_all()`. Since I'm also working on my implementation, I'll test this function myself.
Documentation for sync_all(): [Rust File::sync_all()](https://doc.rust-lang.org/std/fs/struct.File.html#method.sync_all)

Reply:
@harry-cluml We should not mention private repos. https://github.com/aicers/crusher/pull/118#discussion_r1846183790 Could you remove it from your comment?

Reply:
@sophie-cluml I'm truly sorry about that. I've made the corrections now.